Fix Jenkins virtualenv issue that masked stale dependency

timmc-edx opened the pull request:
This fixes a dependency management issue on Jenkins by using `pip-sync` rather than `pip install`, since the latter leaves stale dependencies in the copy of the cached virtualenv.

feanil commented on scripts/xdist/setup_worker.sh:
```diff
@@ -0,0 +1,27 @@
+sed -i "s|\(^/home/jenkins\)/shallow-clone|\1/edx-platform|" -- \
+    $venv/lib/python*/site-packages/*.egg-link
+pip install -qr requirements/edx/pip-tools.txt
+pip-sync requirements/edx/testing.txt "${DJANGO_REQUIREMENT}"
```
@jmbowman I feel like there was a reason why we don't do this but also tests seem to be passing.  Do you remember the reason?

robrap replied:
I think this would be ideal for fixing our current issue, if there are no downsides.  Thanks @timmc-edx.

jmbowman replied:
@jinder1s was the last person to try this in edx-platform: https://github.com/edx/edx-platform/pull/22315 .  Maybe the egg-link hack fixes the issue he encountered?

jmbowman commented on scripts/xdist/setup_worker.sh:
```diff
@@ -0,0 +1,27 @@
+venv=$venv_parent/edx-venv
+rm -rf $venv
+tar -C $venv_parent -xf /home/jenkins/edx-venv_clean-${PYTHON_VERSION}.tar.gz
+source $venv/bin/activate
```
I can never remember if the environment variables initialized this way actually propagate to the environment that called this script (they need to here).

timmc-edx replied:
Ah right, because this is now being sourced from a script. Hmm... it looks like maybe some of our code in paver is actually taking care of activating the venv before running the xdist workers: https://github.com/edx/edx-platform/blob/5de9eed420f0e84db1383403f932e48e98c89df8/pavelib/utils/test/suites/pytest_suite.py#L321 (and on another line as well)

timmc-edx commented:
Where does the main virtualenv get initialized? I see a venv activate in jenkins-common.sh, but nothing that does a pip install there. I guess it relies on the tarball being accurate.

...and I see that edx-platform's own Makefile also doesn't use pip-sync—I should switch `make requirements` to use pip-sync as well. Just running it locally, it looks like both `pip install` and `pip-sync` go to all the work of uninstalling and reinstalling all the local packages, just in a different order so I don't think there's any drastic difference. (`pip-sync` does all the uninstallations first and then the reinstallations, whereas `pip install` interleaves them package by package.)

jmbowman commented:
For normal test runs, `jenkins-common.sh` activates the virtualenv and the `install_python_prereqs` paver task updates it (which means that we do rely on the venv snapshot at least being able to run paver correctly).  For runs that use tox to test an alternate Django or Python version, a tox virtualenv is created and/or activated by https://github.com/edx/edx-platform/blob/master/scripts/generic-ci-tests.sh#L73 .

timmc-edx commented:
I've rebased, since another branch has since fixed the problematic test anyhow. And I added another commit to convert the other main usages of the virtualenv to use pip-sync, and to restore the quiet flag.

jinder1s commented:
When I was trying this, pip-sync did not know how to handle the "-e ." in base.txt(from local.in), that no longer seems to be a problem. So I think this is good to go.

timmc-edx commented:
OK, cool! Squashing the latter two commits together before merging.

timmc-edx commented:
jenkins run python
jenkins run quality

timmc-edx commented:
jenkins run python
